Bluetooth: audio: ascs: Fix missing ASCS/ASE cleanup

MariuszSkamra opened the pull request:
This adds fixes for ASE/ASCS cleanup that was missing in few places.
The unit tests have been added that trigger invalid behavior fixed in this PR.

Fixes: #56111
Fixes: #56138
Fixes: #56139
Fixes: #51552

MariuszSkamra committed:
Bluetooth: audio: ascs: Cleanup ASCS on callback unregister

This adds cleanup of ASCS that is called when upper layer unregisters
it's callbacks. Without callbacks registered, the service is unusable.
The ASE's that were in non-idle state stay in this state, because the
implementation returns an error on every operation that is performed by
the client. The cleanup added moves all the ASE's to idle state.

Fixes: #56111
Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: ascs: Remove redundant check

The ID is ensured to be valid before ase_new() function is called, as
the client request is already validated.
On the other hand, if function is called from server initiated procedure
context, the assert shall occur as it indicates stack issue.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: ascs: Minor refactor of ASE allocation

This removes redundant bt_ascs_ase_get_from_slab function and moves it's
functionality to ase_new(). It's sane, because the ASE shall be
allocated from ase_new() function context only.
bt_ascs_ase_return_to_slab() has been renamed to ase_free() to follow
the naming convention. Furthermore, few logs have been fixed/added.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: ascs: Fix ASE leak if codec config failed

This fixes ASE that is not returned to pool when Codec Config
operation failed. The ASE is allocated in Codec Config operation
context. With this fix, if the operation failed to complete,
the ASE is returned back to the pool, so that it can be allocated
again by this or other client.

Fixes: #56138
Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: ascs: Fix Response_Code for mandatory opcodes

This fixes invalid Unsupported Opcode Response_Code that
happened to be sent in ASE Control Point notification
as the response to operation initiated with Mandatory opcode
to support.
When the Unicast Server callbacks are not registered, every
operation that is initiated by client will fail with
Unspecified Error instead.

Fixes: #56139
Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: iso: Add bt_iso_server_unregister function

This adds function that unregisters previously registered ISO server.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: ascs: Register ISO server on ASCS init

This moves the ISO server registration to bt_ascs_init.
When the ASCS gets cleaned up (the callbacks are unregistered), the ISO
server gets unregistered.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: ascs: Fix ISO server listen incorrect behavior

This fixes incorrect behavior of ISO server listen that was not handling
the case of two endpoints sharing the same CIS/CIG and the endpoints
belonging to different connections.

Fixes: #51552
Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -71,6 +71,7 @@ K_MEM_SLAB_DEFINE(ase_slab, sizeof(struct bt_ascs_ase),
 static struct bt_ascs sessions[CONFIG_BT_MAX_CONN];
 NET_BUF_SIMPLE_DEFINE_STATIC(ase_buf, CONFIG_BT_L2CAP_TX_MTU);
 static struct bt_bap_stream *enabling[CONFIG_BT_ISO_MAX_CHAN];
+static const struct bt_bap_unicast_server_cb *unicast_server_cb;
```
I do not agree it creates stronger coupling.
The callback structure can be easily replaced now with ascs-type callback (I mean something like `bt_ascs_cb`).

> An alternative could be to have a getting function if the `extern` is the issue, e.g.
> 
> ```c
> if (IS_ENABLED(CONFIG_BT_BAP_UNICAST_SERVER) {
>      bt_bap_unicast_server_cb_get();
> }
> ```
I would say that this approach would create stronger coupling :D

Thalley replied:
Fair - Agree to disagree :D 

In any case, there is a strong coupling right now, and since this doesn't really change that, it's fine :)

MariuszSkamra committed:
Bluetooth: audio: ascs: Make use of SYS_SLIST_FOR_EACH_CONTAINER_* macros

This cleans up the code a bit by replacing the occurences of
SYS_SLIST_FOR_EACH_NODE with SYS_SLIST_FOR_EACH_CONTAINER(_SAFE)
so that the explicit CONTAINER_OF() usage is not needed.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: bap_stream: Remove reference to conn internals

This removes the usage of conn internal API in bap_stream.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: Fix declaring conn type object in header file

The fact of having such extern in iso_internal.h forces the source files
that include this header to include conn_internal.h. This breaks the
encapsulation of conn object.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: ascs: Fix double bt_bap_stream_detach call

This fixes possible double bt_bap_stream_detach() function call.
The bt_bap_stream_detach() called from ascs_ep_set_state
state BT_BAP_EP_STATE_IDLE handler.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: ascs: Unbind endpoint in idle state

This moves duplicated bt_bap_iso_unbind_ep function call to
be handled when ASE enters BT_BAP_EP_STATE_IDLE state.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1001,19 +1001,14 @@ static void ase_disable(struct bt_ascs_ase *ase)
-
-		ase = CONTAINER_OF(ase_node, struct bt_ascs_ase, node);
-		stream = ase->ep.stream;
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&session->ases, ase, tmp, node) {
+		struct bt_bap_stream *stream = ase->ep.stream;
```
Am I missing something here? 
The _SAFE macros make sense to use when removing something from the list, but from what I can tell, the ASES are not being removed from `session` here, so can't we just just the non-SAFE version?

Thalley replied:
Ah, yes, because `ase_release -> ascs_ep_set_state(BT_BAP_EP_STATE_RELEASING) -> ascs_ep_set_state(ep, BT_BAP_EP_STATE_IDLE) -> ase_free` 

Given the fairly long list of function calls here (including a recursive one) where the change to `BT_BAP_EP_STATE_IDLE` is conditional, I think we should add a comment here :)

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1636,12 +1629,10 @@ static ssize_t ascs_config(struct bt_ascs *ascs, struct net_buf_simple *buf)
-	SYS_SLIST_FOR_EACH_NODE(&ascs->ases, ase_node) {
-		struct bt_ascs_ase *ase = CONTAINER_OF(ase_node, struct bt_ascs_ase, node);
-
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&ascs->ases, ase, tmp, node) {
```
I assume that the _SAFE version is chosen here in case that the callback removes an ASE?

MariuszSkamra replied:
As it's used inside of `bt_ascs_foreach_ep`, one cannot say that it won't be used to remove the entries from the list - eg. upper layer will call `bt_bap_unicast_server_release` from `bt_bap_ep_func_t` context. Thus it's better to use _SAFE version of the iterator.

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -558,55 +568,41 @@ static int ascs_ep_get_status(struct bt_bap_ep *ep, struct net_buf_simple *buf)
 
-		if (enabling[i] == NULL && free_stream == NULL) {
-			free_stream = &enabling[i];
+		chan = &ase->ep.iso->chan;
```
Can we be sure that `ase->ep.iso` is not NULL here?

MariuszSkamra replied:
I see that there was no check for ISO not being NULL before my change.
Looking int the code, the ISO gets assigned in `ase_stream_qos` function context. Thus I would say the pointer is valid.
But I consider adding assert here.

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -558,55 +568,41 @@ static int ascs_ep_get_status(struct bt_bap_ep *ep, struct net_buf_simple *buf)
+		chan = &ase->ep.iso->chan;
+		if (chan->iso != NULL) {
+			LOG_WRN("ase %p chan %p already connected", ase, chan);
+			break;
```
Should this return `EALREADY` instead of `EPERM` then?

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -558,55 +568,41 @@ static int ascs_ep_get_status(struct bt_bap_ep *ep, struct net_buf_simple *buf)
-	LOG_ERR("Unable to listen: no slot left");
-
-	return -ENOSPC;
+	return -EPERM;
```
Hmm, in libc `EPERM` is defined as
```c
#define EPERM 1         /**< Not owner */
```

Perhaps `EACCES` is better suited here?
```c
#define EACCES 13       /**< Permission denied */
```

MariuszSkamra replied:
The return code was already there, did not changed it. But I agree, the `EACCES` would be better.

Thalley commented on subsys/bluetooth/host/iso_internal.h:
Not sure that I understand your reasoning. You write 

> The fact of having such extern in iso_internal.h forces the source files that include this header to include conn_internal.h. This breaks the encapsulation of conn object.

Why would a source file that include `iso_internal.h` need to include `conn_internal.h`?

MariuszSkamra replied:
It's because bt_conn object is defined in `conn_internal.h`.
In the code below, the `struct bt_conn` is referenced directly (not by the pointer) so the compiler needs to know it's size. Note that this is an array of `struct bt_conn` objects, not it's pointers.
```
extern struct bt_conn iso_conns[CONFIG_BT_ISO_MAX_CHAN];
```
So having it in `iso_internal.h` forces the code that includes it to include the `conn_internal.h` as well. It is straightforward, as how would you e.g. iterate the `iso_conns` not knowing the `bt_conn` size

Thalley replied:
Ah, that makes sense, yeah.

MariuszSkamra committed:
tests: Bluetooth: audio: Extend ISO mock implementation

This adds ISO mock with very basic functionality that is needed to mock
the `bt_iso_` functions used by ASCS.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: Bluetooth: audio: Extend kernel mocks

This adds basic functionality that mocks the k_work API. The ASCS uses
delayed work to defer the CIS disconnection.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: Bluetooth: ascs: Add test for Sink entering streaming state

This adds unit test for Sink ASE entering streaming state.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/audio/bap_stream.c:
```diff
@@ -377,7 +387,7 @@ int bt_bap_stream_enable(struct bt_bap_stream *stream,
 	}
 
-	role = stream->conn->role;
+	role = conn_get_role(stream->conn);
```
Could consider implement a `stream_conn_get_role(stream)` too
